### ISSUE TITLE
PUK vs GPG Admin PIN

### ISSUE THREAD

**Issue by andrewpmartinez** [drduh/YubiKey-Guide#248]
With a brand new Yubikey 5 Nano, I changed my PIN and PUK via the Yubikey managers. However, my GPG Admin PIN was the default 12345678 and is independent of the PIV PIN/PIV PUK.

It appears that the section on PINs (https://github.com/drduh/YubiKey-Guide#configure-smartcard) conflates the GPG Admin PIN and the PUK (pin unblock key).

> The default PIN is 123456 and default Admin PIN (PUK) is 12345678

Where "default Admin PIN (PUK)" leads me to believe the Admin PIN == PUK.

From my testing, the following appears to be the case using GPG

-  GPG PIN
   - Defaults to 123456
   - can be 6-8 characters
   - can be changed via `gpg --card-edit` -> `passwd` w/o admin settings
   - can be changed via `gpg --card-edit` -> `admin` -> `passwd` -> `1`  w/ admin settings on
-  GPG PUK is used to reset the wrong PIN entry count
   - Defaults to 123456
   - can be 6-8 characters
   - can be changed via `gpg --card-edit` -> `admin` -> `passwd` -> `2`  w/ admin settings on
- GPG Admin PIN
   - Defaults to 12345678, must be 8 characters
   - can be changed via `gpg --card-edit` -> `admin` -> `passwd` -> `3`  w/ admin settings on

There are related commands available via `ykman`: https://docs.yubico.com/ykman/OpenPGP_Commands.html

If this is actually correct I don't mind submitting a PR, I want to confirm what I am seeing and my understanding.

Edit: Updated to reflect PIV vs GPG module

**Comment by javasboy** [comment]
I have the same doubts, now I only know the GPG pin and admin pin.

**Comment by andrewpmartinez** [comment]
I was reading through the Yubikey documentation and the GPG module is 100% separate from the PIV module. Separate PIN, PIN Unblock, and Admin PIN.

**Comment by andrewpmartinez** [comment]
I just popped in a spare Yubikey I have from work for testing and the PIV and GPG modules are 100% separate, but share similar defaults. I updated my original comment. The article should be updated to include the following:

- PIV and GPG modules have a separate PIN, PUK, and reset functionality
- The GPG module has an additional Admin PIN that is defaulted to 12345678 and must be 8 characters. Prefernce is put upon numerical PINs as some software/HSMs don't deal with alpha character well
- The GPG Module is not configurable via the Yubikey Manager but it can be configured via `.\ykman.exe openpgp` and `gpg --card-edit`

**Comment by Paraphraser** [comment]
On the topic of PINs, something I noticed early on and didn't think to comment on until now occurs on my iMac where I have an extended keyboard with numeric pad. I can't enter a PIN using the numeric pad. I. The only way to unlock a YubiKey is to enter the PIN on the top strip.

I don't know if this is a macOS thing, the particular kind of keyboard thing (Apple wireless) or applies to all OSes, all keyboards wired and wireless.

Let's just say it used up two of my three tries before I thought, "I might've made a mistake on the first entry, I know I was careful on the second try, what's going on? Maybe try the top strip. Ahah!"

**Comment by andrewpmartinez** [comment]
@Paraphraser  I can't find the link, but I was reading a resource on GPG PINs and it suggested only using numeric-only PINs due to some keyboard mapping problems. I wish I had the link, but I blew through it looking for other answers. I originally thought it meant non-English keyboards but apparently, it might affect any non-Latin based pc keyboard layout.

Might be worth a note, but I haven't dug in enough to feel like I can write it. If someone else wants to research it and setup another PR I'd suggest that.

**Comment by lleddewk** [comment]
My reading of this is slightly different:

1.  I agree that PIV and Openpgp modes are completely separate and the PIV PUK has no effect on the Opengpg mode.  Re-setting the PUK is therefore a red-herring.

2.  The GPG PUK is called a Reset Code (see attached sample gpg --card-edit output) and can be set using gpg --card edit -> admin -> passwd.

3.  By default the retry count for the Reset Code is zero (at least on a yubikey 4 or 5). The code is therefore unusable and there is no need to change from default.  See attached sample output.

4. The maximum length of each of the PINs is 127 characters.  Again see sample output.  I am using a 12 character PIN so the maximum length of 8 is not correct.

Comments?


Keith

[card-status.txt](https://github.com/drduh/YubiKey-Guide/files/6682030/card-status.txt)

**Comment by drduh** [comment]
All, can you please review @pedrohdz merged commit and comment if that clarifies the ambiguity? Please reopen if there's work remaining.
